packages/cozy-realtime/src/RetryManager.js

```diff
@@ -0,0 +1,142 @@
+import MicroEE from 'microee'
+
+import {
+  maxWaitBetweenRetries,
+  baseWaitAfterFirstFailure,
+  timeBeforeSuccessfull,
+  raiseErrorAfterAttempts
+} from './config'
+
+class RetryManager {
+  constructor() {
+    this.reset()
+    this.onSuccess = this.onSuccess.bind(this)
+    this.onFailure = this.onFailure.bind(this)
+    this.onSuccessAcknowledged = this.onSuccessAcknowledged.bind(this)
+  }
+
+  /**
+   * On a success
+   *
+   * If `timeBeforeSuccessfull` if not null, it will wait this amount of
+   * time without errors before acknowleding a success
+   */
+  onSuccess() {
+    if (timeBeforeSuccessfull > 0) {
+      this.startSuccessWatcher(
+        this.onSuccessAcknowledged,
+        timeBeforeSuccessfull
+      )
+    } else {
+      this.onSuccessAcknowledged()
+    }
+  }
+
+  /**
+   * Acknowledge a success
+   *
+   * @private
+   */
+  onSuccessAcknowledged() {
+    this.emit('success')
+    this.reset()
+  }
+
+  /**
+   * Clears the success watcher
+   *
+   * @private
+   */
+  clearSuccessWatcher() {
+    if (this.successWatcher) {
+      global.clearTimeout(this.successWatcher)
+      this.successWatcher = null
+    }
+  }
+
+  /**
+   * Starts the success watcher - acknowledges the success if not interrupted after some time
+   *
+   * @private
+   * @param {function} ack - will be called after the wait time
+   * @param {number} time - waiting time
+   */
+  startSuccessWatcher(ack, time) {
+    if (!this.successWatcher) {
+      const callback = () => {
+        this.successWatcher = null
+        ack()
+      }
+      this.successWatcher = global.setTimeout(callback, time)
+    }
+  }
+
+  /**
+   * On failure
+   *
+   * If we where waiting for a success timeout, it will cancel the wait.
+   *
+   * An application error (failed authentication, failed subscription) will
+   * not trigger a failure (trying to reconnect with the same parameters
+   * has no meaning anyways).
+   *
+   * @param {object} error
+   */
+  onFailure(error) {
+    this.clearSuccessWatcher()
+    this.increaseFailureCounter()
+    this.emit('failure', error)
+    if (this.hasReachedErrorLevel()) this.emit('error', error)
+  }
+
+  /**
+   * Returns if we reached a level where we need to raise an error
+   *
+   * @returns {boolean}
+   */
+  hasReachedErrorLevel() {
+    return this.retries > 0 && this.retries % raiseErrorAfterAttempts == 0
+  }
+
+  /**
+   * Reset failures counters (do not wait for before trying to connect next time)
+   */
+  reset() {
+    this.clearSuccessWatcher()
+    this.retries = 0
+    this.wait = false
+  }
+
+  /**
+   * Increase the failure counter (wait more before next attempt)
+   *
+   * @private
+   */
+  increaseFailureCounter() {
+    if (this.wait === false) {
+      // first reconnect is immediate
+      this.wait = 0
```

Comment:
What is the difference between `this.wait = 0` and `thiss.wait = false`

Reply:
This is an ugly hack to have the first reconnect to be immediate. 
false (no wait) -> 0 (wait 0ms) -> 128 (wait 128ms) -> …

I am adding comments for  this